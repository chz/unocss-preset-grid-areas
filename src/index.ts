import type { Preset } from "unocss"

export interface GridOptions {
  gridTemplateAreas: {
    [key: string]: string[];
  };
}

export function presetGridAreas(options: GridOptions): Preset {
   const gridTemplateAreas = (name) => options?.gridTemplateAreas?.[name] ? options?.gridTemplateAreas?.[name].map((row) => (`"${row}"`)).join('\n') : ''

  return {
    name: "unocss-preset-grid-areas",
    rules: [
      ['grid-areas-none', { 'grid-template-areas': 'none' }],
      ['grid-areas-inherit', { 'grid-template-areas': 'inherit' }],
      ['grid-areas-initial', { 'grid-template-areas': 'initial' }],
      ['grid-areas-revert', { 'grid-template-areas': 'revert' }],
      ['grid-areas-revert-layer', { 'grid-template-areas': 'revert-layer' }],
      ['grid-areas-unset', { 'grid-template-areas': 'unset' }],
      ['grid-in-auto', { 'grid-area': 'auto' }],
      ['grid-in-inherit', { 'grid-area': 'inherit' }],
      ['grid-in-initial', { 'grid-area': 'initial' }],
      ['grid-in-revert', { 'grid-area': 'revert' }],
      ['grid-in-revert-layer', { 'grid-area': 'revert-layer' }],
      ['grid-in-unset', { 'grid-area': 'unset' }],
      [/^grid-in-(.+)$/, ([_, name]) => ({ 'grid-area': `${name}` })],
      [/^row-start-(.+)$/, ([_, name]) => ({ 'grid-row-start': `${name}-start` })],
      [/^row-end-(.+)$/, ([_, name]) => ({ 'grid-row-end': `${name}-end` })],
      [/^col-start-(.+)$/, ([_, name]) => ({ 'grid-column-start': `${name}-start` })],
      [/^col-end-(.+)$/, ([_, name]) => ({ 'grid-column-end': `${name}-end` })],
      [/^grid-areas-(.*)$/, ([_, name]) => ({ 'grid-template-areas': gridTemplateAreas(name) })],
    ],
  }
}

export default presetGridAreas
