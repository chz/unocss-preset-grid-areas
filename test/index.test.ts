import { createGenerator, presetAttributify, presetUno } from 'unocss'
import { describe, expect, it } from 'vitest'
import { createAutocomplete } from '@unocss/autocomplete'
import { presetGridAreas } from '../src'

describe('grid-areas', () => {
  const generator = createGenerator({
    presets: [
      presetUno(),
      presetAttributify(),
      presetGridAreas({
        gridTemplateAreas: {
          'layout': [
            'header header header',
            'nav    main   main',
            'nav    footer footer',
          ],
        },
      }),
    ],
  })

  const ac = createAutocomplete(generator)

  async function enumerateSuggestions(inputs: string[]) {
    return Object.fromEntries(await Promise.all(inputs.map(async input => [
      input,
      (await ac.suggest(input)).slice(0, 10).join(' '),
    ])))
  }

  it('utilities', async () => {
    const { css } = await generator.generate([
      'grid-areas-layout',
      'grid-in-footer',
      'grid-in-header',
      'grid-in-main',
      'grid-in-nav',
      'col-start-footer',
      'col-start-header',
      'col-start-main',
      'col-start-nav',
      'col-end-footer',
      'col-end-header',
      'col-end-main',
      'col-end-nav',
      'row-start-footer',
      'row-start-header',
      'row-start-main',
      'row-start-nav',
      'row-end-footer',
      'row-end-header',
      'row-end-main',
      'row-end-nav',
    ].join(' '))

    expect(css).toMatchSnapshot()
  })

  it('global keys', async ()=>{
    const { css } = await generator.generate([
      'grid-areas-none',
      'grid-areas-inherit',
      'grid-areas-initial',
      'grid-areas-revert',
      'grid-areas-revert-layer',
      'grid-areas-unset',
      'grid-in-auto',
      'grid-in-inherit',
      'grid-in-initial',
      'grid-in-revert',
      'grid-in-revert-layer',
      'grid-in-unset',
    ].join(' '))

    expect(css).toMatchSnapshot()
  })
})
