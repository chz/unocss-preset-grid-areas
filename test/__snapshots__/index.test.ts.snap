// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`grid-areas > global keys 1`] = `
"/* layer: preflights */
*,::before,::after{--un-rotate:0;--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-scale-x:1;--un-scale-y:1;--un-scale-z:1;--un-skew-x:0;--un-skew-y:0;--un-translate-x:0;--un-translate-y:0;--un-translate-z:0;--un-pan-x: ;--un-pan-y: ;--un-pinch-zoom: ;--un-scroll-snap-strictness:proximity;--un-ordinal: ;--un-slashed-zero: ;--un-numeric-figure: ;--un-numeric-spacing: ;--un-numeric-fraction: ;--un-border-spacing-x:0;--un-border-spacing-y:0;--un-ring-offset-shadow:0 0 rgba(0,0,0,0);--un-ring-shadow:0 0 rgba(0,0,0,0);--un-shadow-inset: ;--un-shadow:0 0 rgba(0,0,0,0);--un-ring-inset: ;--un-ring-offset-width:0px;--un-ring-offset-color:#fff;--un-ring-width:0px;--un-ring-color:rgba(147,197,253,0.5);--un-blur: ;--un-brightness: ;--un-contrast: ;--un-drop-shadow: ;--un-grayscale: ;--un-hue-rotate: ;--un-invert: ;--un-saturate: ;--un-sepia: ;--un-backdrop-blur: ;--un-backdrop-brightness: ;--un-backdrop-contrast: ;--un-backdrop-grayscale: ;--un-backdrop-hue-rotate: ;--un-backdrop-invert: ;--un-backdrop-opacity: ;--un-backdrop-saturate: ;--un-backdrop-sepia: ;}::backdrop{--un-rotate:0;--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-scale-x:1;--un-scale-y:1;--un-scale-z:1;--un-skew-x:0;--un-skew-y:0;--un-translate-x:0;--un-translate-y:0;--un-translate-z:0;--un-pan-x: ;--un-pan-y: ;--un-pinch-zoom: ;--un-scroll-snap-strictness:proximity;--un-ordinal: ;--un-slashed-zero: ;--un-numeric-figure: ;--un-numeric-spacing: ;--un-numeric-fraction: ;--un-border-spacing-x:0;--un-border-spacing-y:0;--un-ring-offset-shadow:0 0 rgba(0,0,0,0);--un-ring-shadow:0 0 rgba(0,0,0,0);--un-shadow-inset: ;--un-shadow:0 0 rgba(0,0,0,0);--un-ring-inset: ;--un-ring-offset-width:0px;--un-ring-offset-color:#fff;--un-ring-width:0px;--un-ring-color:rgba(147,197,253,0.5);--un-blur: ;--un-brightness: ;--un-contrast: ;--un-drop-shadow: ;--un-grayscale: ;--un-hue-rotate: ;--un-invert: ;--un-saturate: ;--un-sepia: ;--un-backdrop-blur: ;--un-backdrop-brightness: ;--un-backdrop-contrast: ;--un-backdrop-grayscale: ;--un-backdrop-hue-rotate: ;--un-backdrop-invert: ;--un-backdrop-opacity: ;--un-backdrop-saturate: ;--un-backdrop-sepia: ;}
/* layer: default */
.grid-areas-none{grid-template-areas:none;}
.grid-areas-inherit{grid-template-areas:inherit;}
.grid-areas-initial{grid-template-areas:initial;}
.grid-areas-revert{grid-template-areas:revert;}
.grid-areas-revert-layer{grid-template-areas:revert-layer;}
.grid-areas-unset{grid-template-areas:unset;}
.grid-in-auto{grid-area:auto;}
.grid-in-inherit{grid-area:inherit;}
.grid-in-initial{grid-area:initial;}
.grid-in-revert{grid-area:revert;}
.grid-in-revert-layer{grid-area:revert-layer;}
.grid-in-unset{grid-area:unset;}"
`;

exports[`grid-areas > utilities 1`] = `
"/* layer: preflights */
*,::before,::after{--un-rotate:0;--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-scale-x:1;--un-scale-y:1;--un-scale-z:1;--un-skew-x:0;--un-skew-y:0;--un-translate-x:0;--un-translate-y:0;--un-translate-z:0;--un-pan-x: ;--un-pan-y: ;--un-pinch-zoom: ;--un-scroll-snap-strictness:proximity;--un-ordinal: ;--un-slashed-zero: ;--un-numeric-figure: ;--un-numeric-spacing: ;--un-numeric-fraction: ;--un-border-spacing-x:0;--un-border-spacing-y:0;--un-ring-offset-shadow:0 0 rgba(0,0,0,0);--un-ring-shadow:0 0 rgba(0,0,0,0);--un-shadow-inset: ;--un-shadow:0 0 rgba(0,0,0,0);--un-ring-inset: ;--un-ring-offset-width:0px;--un-ring-offset-color:#fff;--un-ring-width:0px;--un-ring-color:rgba(147,197,253,0.5);--un-blur: ;--un-brightness: ;--un-contrast: ;--un-drop-shadow: ;--un-grayscale: ;--un-hue-rotate: ;--un-invert: ;--un-saturate: ;--un-sepia: ;--un-backdrop-blur: ;--un-backdrop-brightness: ;--un-backdrop-contrast: ;--un-backdrop-grayscale: ;--un-backdrop-hue-rotate: ;--un-backdrop-invert: ;--un-backdrop-opacity: ;--un-backdrop-saturate: ;--un-backdrop-sepia: ;}::backdrop{--un-rotate:0;--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-scale-x:1;--un-scale-y:1;--un-scale-z:1;--un-skew-x:0;--un-skew-y:0;--un-translate-x:0;--un-translate-y:0;--un-translate-z:0;--un-pan-x: ;--un-pan-y: ;--un-pinch-zoom: ;--un-scroll-snap-strictness:proximity;--un-ordinal: ;--un-slashed-zero: ;--un-numeric-figure: ;--un-numeric-spacing: ;--un-numeric-fraction: ;--un-border-spacing-x:0;--un-border-spacing-y:0;--un-ring-offset-shadow:0 0 rgba(0,0,0,0);--un-ring-shadow:0 0 rgba(0,0,0,0);--un-shadow-inset: ;--un-shadow:0 0 rgba(0,0,0,0);--un-ring-inset: ;--un-ring-offset-width:0px;--un-ring-offset-color:#fff;--un-ring-width:0px;--un-ring-color:rgba(147,197,253,0.5);--un-blur: ;--un-brightness: ;--un-contrast: ;--un-drop-shadow: ;--un-grayscale: ;--un-hue-rotate: ;--un-invert: ;--un-saturate: ;--un-sepia: ;--un-backdrop-blur: ;--un-backdrop-brightness: ;--un-backdrop-contrast: ;--un-backdrop-grayscale: ;--un-backdrop-hue-rotate: ;--un-backdrop-invert: ;--un-backdrop-opacity: ;--un-backdrop-saturate: ;--un-backdrop-sepia: ;}
/* layer: default */
.grid-in-footer{grid-area:footer;}
.grid-in-header{grid-area:header;}
.grid-in-main{grid-area:main;}
.grid-in-nav{grid-area:nav;}
.row-start-footer{grid-row-start:footer-start;}
.row-start-header{grid-row-start:header-start;}
.row-start-main{grid-row-start:main-start;}
.row-start-nav{grid-row-start:nav-start;}
.row-end-footer{grid-row-end:footer-end;}
.row-end-header{grid-row-end:header-end;}
.row-end-main{grid-row-end:main-end;}
.row-end-nav{grid-row-end:nav-end;}
.col-start-footer{grid-column-start:footer-start;}
.col-start-header{grid-column-start:header-start;}
.col-start-main{grid-column-start:main-start;}
.col-start-nav{grid-column-start:nav-start;}
.col-end-footer{grid-column-end:footer-end;}
.col-end-header{grid-column-end:header-end;}
.col-end-main{grid-column-end:main-end;}
.col-end-nav{grid-column-end:nav-end;}
.grid-areas-layout{grid-template-areas:\\"header header header\\"
\\"nav    main   main\\"
\\"nav    footer footer\\";}"
`;
